Add Discord area sub-regions, refactor areas into GameArea enum and use for other plugins' region lookups

Nightfirecat opened the pull request:
Fixes #5364
Fixes #10236

The primary goals of this PR are as follows:

1. Add support for sub-region matching in the Discord plugin (ex. White Wolf Mountain, separate gwd bosses from the gwd regions, maps which are different on separate planes (cam torum/neypotzli, kbd/nmz, mta rooms))
2. Split area definitions out from `DiscordGameEventType`. This makes their definitions public and usable in the client, as well as in hub plugins. (this creates incentive for hub plugin authors to proactively contribute updates for a maintained region ID reference, while also keeps the Discord plugin's defined regions up-to-date)
3. Improve area enum naming; because of area definitions being made public as part of (2), the previous prefixed enum names are very clunky and generally unhelpful when being referenced from other classes/plugins.

<details><summary>Draft -> Ready-for-Review Changelog</summary>
<ul>
<li>a5c552c -> 2660b46: Add <code>GameArea#contains(WorldPoint)</code> and <code>RegionArea(int, WorldPoint, WorldPoint)</code></li>
<li>2660b46 -> 6852c69: Fix timers plugin test to not fail due to implementation change, fix DiscordState to not break test anymore (and add further tests to ensure coverage), remove unused imports</li>
<li>6852c69 -> a4998b0: Fix PMD violation (unused private method in loot tracker pulgin)</li>
<li>a4998b0 -> 2a81e6e: Rebase after <code>client.getVar()</code> -> <code>client.getVarbitValue()</code> API change</li>
<li>2a81e6e -> b8b970e: Remove ugly GameArea name prefixes. This (imo) is the last naming fix this PR needs</li>
<li>b8b970e -> e4f4933: Finalize commits for non-draft review</li>
<li>e4f4933 -> 89c7f62: Many rebases, adding support for planar <code>RegionArea</code>s</li>
<li>89c7f62 -> 79c31a2: GameArea fixes (renames, combining/splitting areas, improvements/fixes to sub-region areas)</li>
<li>79c31a2 -> e0c084e: Remove unused nmz region var, more GameArea fixes and expansions</li>
</ul>
</details>

Nightfirecat committed:
WorldArea: add lombok, add intersects2D

LlemonDuck commented on runelite-client/src/main/java/net/runelite/client/game/GameArea.java:
```diff
@@ -0,0 +1,531 @@
+	BOSS_COMMANDER_ZILYANA("Commander Zilyana", GameAreaType.BOSSES, new RegionArea(11602, new WorldArea(2889, 5258, 19, 18, 0))),
+	BOSS_CORP("Corporeal Beast", GameAreaType.BOSSES, 11842, 11844),
+	BOSS_DKS("Dagannoth Kings", GameAreaType.BOSSES, 11588, 11589),
+	BOSS_GENERAL_GRAARDOR("General Graardor", GameAreaType.BOSSES, new RegionArea(11347, new WorldArea(2864, 5351, 13, 19, 2))),
```
Will the `WorldArea`s within this work in instanced versions of these regions? e.g. Ironman instance of Bandos here

Nightfirecat replied:
No, they are drawn for the non-instanced map. `DiscordGameEventType.fromPoint(WorldPoint)` expects to be given a point adjusted for instances. (eg. from `WorldPoint.fromLocalInstance(Client, LocalPoint)`)

Nightfirecat commented:
I've finished all of the to-do items from the draft version of this PR and marked it as ready for review. I've been running a dev build with this branch merged for the past week or so and can confirm that it is working the same as release generally, and that sub-region recognition is working as expected for overworld locations. (have not yet confirmed instance sub-region areas)

Nightfirecat commented:
I have found one bug with this branch thus far: when running Theatre of Blood, buying/obtaining items from the supply chests after bloat and Sotetseg rooms incorrectly triggers chest loot detection. I will look into this and fix it.

Nightfirecat commented:
I've fixed the Theatre of Blood chest issue by creating a separate area for Verzik Vitur's treasure vault and using that as the area to check for chest loot rather than the full Theatre of Blood area.

Nightfirecat commented:
I've identified a bug with this PR--if you enter an area which is not mapped, skill activity updates are not tracked and not displayed in rich presence. I'll look at addressing that this weekend.

Nightfirecat committed:
discord: add unknown area and skilling change test

Nightfirecat commented:
I've corrected this by removing an errant `gameArea == null` guard in the plugin, and have added a test which demonstrates its behavior working.

A small note: this PR does change the behavior of area changes slightly--previously, moving into a new area would reset your current skill activity. With this PR applied, your trained skill is maintained through area changes. (and even to `IN_GAME` ie. unknown areas)

Nightfirecat committed:
game: split GameArea from Discord event types

Nightfirecat committed:
GameArea: improve enum names, fix spellings

Some entries have also been combined, split apart, or otherwise renamed
accordingly. (Entries now generally line up one-to-one with wiki pages
so as not to be difficult to search for.)

Nightfirecat committed:
GameArea: support sub-region areas

This commit additionally fixes up, combines, rearranges, or splits up
areas where it makes sense to do so. Some changes were made by
identifying overly large areas due to the previous limitations of only
being able to target full regions, and others were made due to finding
poorly-named or confusing areas.

Nightfirecat committed:
client: use GameArea for region checks

Nightfirecat committed:
improve edgeville dungeon & varrock sewers areas
